chalicelib_smaht/checks/helpers/google_utils.py

```diff
@@ -0,0 +1,1107 @@
+    If no access keys are provided, initiates a connection to production.
+
+    Interfaces with Google services using Google Analytics Data API v1 (GA4).
+
+    For testing against localhost, please provide a `ff_access_keys` dictionary with server=localhost:8000 and key & secret from there as well.
+
+    Arguments:
+        ff_access_keys      - Optional. A dictionary with a 'key', 'secret', and 'server', identifying admin account access keys and FF server to POST to.
+        google_api_key      - Optional. Override default API key for accessing Google.
+        s3UtilsInstance     - Optional. Provide an S3Utils class instance connected to a bucket with a proper Google API key (if none supplied otherwise).
+                              If not supplied, a new S3 connection will be created to the SMaHT production bucket.
+        extra_config        - Additional Google API config, e.g. OAuth2 scopes and Analytics View ID. Shouldn't need to set this.
+    """
+
+    DEFAULT_CONFIG = DEFAULT_GOOGLE_API_CONFIG
+
+    @staticmethod
+    def validate_api_key_format(json_api_key):
+        try:
+            assert json_api_key is not None
+            assert isinstance(json_api_key, dict)
+            assert json_api_key['type'] == 'service_account'
+            assert json_api_key["project_id"].startswith("smaht-")
+            for dict_key in ['private_key_id', 'private_key', 'client_email', 'client_id', 'auth_uri', 'client_x509_cert_url']:
+                assert json_api_key[dict_key]
+        except:
+            return False
+        return True
+
+    def __init__(
+        self,
+        ff_access_keys      = None,
+        google_api_key      = None,
+        s3UtilsInstance     = None,
+        extra_config        = DEFAULT_GOOGLE_API_CONFIG
+    ):
+        """Authenticate with Google APIs and initialize sub-class instances."""
+        if s3UtilsInstance is None:
+            self._s3Utils = s3_utils.s3Utils(env='data')  # Google API Keys are stored on production bucket only ATM.
+        else:
+            self._s3Utils = s3UtilsInstance
+
+        if google_api_key is None:
+            self._api_key = self._s3Utils.get_google_key()
+            if not self._api_key:
+                raise Exception("Failed to get Google API key from S3.")
+        else:
+            self._api_key = google_api_key
+
+        # os.environ['GRPC_DNS_RESOLVER'] = 'native'
+
+        if not GoogleAPISyncer.validate_api_key_format(self._api_key):
+            raise Exception("Google API Key is in invalid format.")
+
+        self.extra_config = extra_config
+        self.credentials = Credentials.from_service_account_info(
+            self._api_key,
+            scopes=self.extra_config.get('scopes', DEFAULT_GOOGLE_API_CONFIG['scopes'])
+        )
+
+        # These are required only for POSTing/GETing data for TrackingInfo items.
+        if ff_access_keys is None:
+            ff_access_keys = self._s3Utils.get_access_keys()
```

Comment:
You want the foursight key no? This will retrive admin keys by default, which may not exist